Dapper/SqlMapper.cs

```diff
@@ -373,39 +385,46 @@ public static void AddTypeHandlerImpl(Type type, ITypeHandler? handler, bool clo
                 }
             }
 
-            var snapshot = typeHandlers;
-            if (snapshot.TryGetValue(type, out var oldValue) && handler == oldValue) return; // nothing to do
+            // synchronize between callers mutating type-handlers; note that regular query
+            // code may still be accessing the field, so we still use snapshot/mutate/swap;
+            // the synchronize is just to prevent lost writes
+            lock (typeHandlersSyncLock)
+            {
+                var snapshot = typeHandlers;
+                if (snapshot.TryGetValue(type, out var oldValue) && handler == oldValue) return; // nothing to do
 
-            var newCopy = clone ? new Dictionary<Type, ITypeHandler>(snapshot) : snapshot;
+                var newCopy = clone ? new Dictionary<Type, ITypeHandler>(snapshot) : snapshot;

```

Comment:
If not using Concurrent, always make a copy, but I think there is still a thread issue during the iteration of the ctor for the `new Dictionary<>`.  If some other code touches the dictionary, which today it might not, but in the future...

Reply:
In a separate push I've suppressed the bad `Impl`'s usage of `clone: false`, so: the only scenario where `clone` is `false` is in the .cctor, which is synchronized (over all field access etc) by the runtime. So the only scenario at real runtime is when we've duplicated *before* mutating. Concurrent reads with zero writes are fine.

"if we add broken code later" is a scenario that we can't plan for other than "try not to do that", but that's always true; we know the dictionary needs concurrency consideration